notes/array.txt

```diff
@@ -0,0 +1,57 @@
+
+
+
+
+
+# defining type k2.Array
+
+
+class Array:
+
+    # `indexes` is a Tensor with one
+    Tensor indexes;
+
+    # `data` is either:
+    #   - of type Tensor (if this corresponds to Array2 == 2-dimensional
+    #     array in C++)
+    #   - of type Array (if this corresponds to Array3 or higher-dimensional
+    #     array in C++)
+    # The Python code is structured a bit differently from the C++ code,
+    # due to the differences in the languages.
+    # When we dispatch things to C++ code there would be some
+    # big switch statement or if-statement to select the right
+    # template instantiation.
+    data;
+
+    def __len__(self):
+        return indexes.shape[0] - 1
+
+    @property
+    def shape(self):
+        # e.g. if indexes.shape is (15,) and
+        # data.shape is (150) -> this.shape would be (15,None)
+        # If data.shape is (150,4), this.shape would be (15,4)
+        # If data.shape is (150,None) (since data is an Array), this.shape
+        #     would be (150,None,None).
+        # The Nones are for dimensions where the shape is not known
+        # because it is variable.
+        return (indexes.shape[0] - 1, None, *data.shape[1:])
```

Comment:
The comments are confusing compared to  
`return (indexes.shape[0] - 1, None, *data.shape[1:])`.

Could you please make it clearer? @danpovey

Reply:
I think indexes.shape should be 16 here.  Can you please edit though?  Sorry I am a bit busy today.  I am hoping @qindazhu can have a look at your PR.  Thanks so much!!